### DISCUSSION TITLE
Embedding Pyright LSP support in new LSP

### DISCUSSION THREAD

**Discussion by thejcannon** [microsoft/pyright#3815] [Q&A]
I'm looking into providing a LSP support for [Pantsbuild](https://www.pantsbuild.org) BUILD files, which are restricted Python (no importing, `exec`'d with a small set of `globals`, etc...).

I'd like to not re-invent the wheel, so ideally I sit this LSP support on top of `pyright`'s existing support. I have a PoC where I can just launch the Pyright LSP (hacking away in VSCode extension), but I'm not really sure where to go from here. Ideally I don't have to repeat much, or curry LSP calls and I could just set the "right" PyRight LSP settings (don't suggest imports, don't go searching for definitions except for one `.pyi` file I can generate providing the "right" definitions, etc...).

If I wanted to go the "extend PyRight's LSP and tweak settings route", I tried:
```js
import { BackgroundAnalysisRunner, run, PyrightServer } from 'pyright/dist/pyright-internal.js';

class PantsBuildServer extends PyrightServer {
    async getSettings(workspace) {
        const serverSettings = {
			//...
        };
        return serverSettings;
    }
}

// ...
```

but `PyrightServer` isn't an exported symbol of `pyright-internal`, so the runtime isn't happy.

I could alternatively go the "tweak settings from the client route", but am unsure how to do this through the LSP (mimicing the client's workspace settings), but still have mapped that one out mentally.

Thanks <3

(Also forgive me, JS/TS is not my strong suit)

**Comment by erictraut** [comment]
Pyright's code wasn't designed for this use case, so you will likely hit issues with this approach. Among other things, there's currently no exposed setting for "don't allow any imports". I wouldn't want to expose such settings in the core pyright code base because this would be a support burden for a very niche use case. You may need to fork the code to do what you want.

If you simply want to extend the global (builtins) namespace, you can use this [existing feature](https://github.com/microsoft/pyright/blob/main/docs/builtins.md).

**Reply by thejcannon** [reply]
In this case, I'm only looking to provide LSP support, mainly for auto-completions, hover support, and light syntax/argument-type checking.

Things like "don't allow imports" are supported well enough by turning off the auto-import suggestion.

If we wen't the `__builtins__.pyi` route, it'd work ok-ish, except we'd be polluting the user's Python LSP support as well :wink: . Which is why I think we need a separate LSP server.

In all honesty, true support would be writing a new LSP somewhat from scratch, but that's a bit out of a volunteer's time.

I ended up vendoring Pyright to do the extension listed above, and plan on just using a git submodule to keep that going longer-term.

**Reply by thejcannon** [reply]
I guess I should say, I'm just looking for some pointers on how to sit on top of Pyright for this case, not looking for anything to change about Pyright. So far I'm looking at a case that:
- Doesn't try to parse any source code for definitions (we only want to LSP `BUILD` files)
- Provides syntax-highlight, autocomplete, red-squiggles, etc.. as if the `BUILD` file was Python
- Allows me to inject some global definitions (should be handled by `__builtins__.pyi` if I can finagle it)

**Reply by r3m0t** [reply]
Did you consider copying this code https://github.com/microsoft/pyright/blob/main/packages/vscode-pyright/src/extension.ts#L94 but using `LanguageClientOptions.uriConverters` to rename the `BUILD` files to `BUILD.py`? Then the LSP would recognise them as Python files. You'd also change the `LanguageClientOptions.documentSelector`.

**Reply by thejcannon** [reply]
That'd be client-side, where my hope is to provide it server-side so other editors can join in.

We also can do this as well with VSCodes `fileAssociations`, but it doesn't lead to a good experience. (it's what I'm doing now so I at least get syntax highlighting)

Lots of red squiggles, pollution with the users Python files (in terms of suggesting funcs/classes) still no easy way to provide the built-in definitions without polluting the python files as well (build targets aren't built-in to user code).
